File: apps/Applications/ios/Local For Vocal/Local For Vocal/Sources/Managers/SpecialDealManager.swift
import Foundation
import SwiftUI

// MARK: - Special Deal Manager
class SpecialDealManager: BasePageManager {
    // Parent Category ID for Special Deals
    override var parentCategoryId: String {
        return "695f88c75f463eeb3c42e765"
    }

    override func fetchInitialData() {
        Task {
            @MainActor in
            self.isLoading = true

            // 1. Fetch Layout from API to get Categories
            do {
                if let layout = try await APIService.shared.fetchLayout(
                    slug: "special-deal-new-style")
                {
                    if let catComponent = layout.components.first(where: {
                        $0.id == "special-categories"
                    }) {
                        // Manually parse items from props
                        if let items = catComponent.props?["items"]?.value as? [[String: Any]] {
                            self.subCategories = items.compactMap { dict in
                                guard let name = dict["name"] as? String else { return nil }
                                let id = dict["id"] as? String ?? UUID().uuidString
                                let image = dict["image_url"] as? String

                                return SubCategory(
                                    id: id, name: name, image: image, icon: nil, slug: nil)
                            }
                        }
                    }
                }
            } catch {
                AppLogger.debug("Error loading Special Deal layout: \(error)")
            }

            // 2. Fetch Products from API
            await self.fetchProducts()
            self.isLoading = false
        }
    }
}

File: apps/Applications/android/app/src/main/java/com/ecommerceearn/app/data/manager/SpecialDealManager.kt
package com.ecommerceearn.app.data.manager

import com.ecommerceearn.app.data.model.Category
import com.ecommerceearn.app.data.remote.NetworkClient
import com.ecommerceearn.app.utils.AppLogger
import kotlinx.coroutines.launch
import java.util.UUID

class SpecialDealManager : BasePageManager() {
    override val parentCategoryId: String
        get() = "695f88c75f463eeb3c42e765"

    override fun fetchInitialData() {
        scope.launch {
            _isLoading.value = true
            
            try {
                val layout = NetworkClient.apiService.getLayoutBySlug("special-deal-new-style")
                val catComponent = layout.components.firstOrNull { it.originalId == "special-categories" }
                
                if (catComponent != null && catComponent.props != null) {
                    val itemsNode = catComponent.props.get("items")
                    if (itemsNode != null && itemsNode.isJsonArray) {
                        val itemsArray = itemsNode.asJsonArray
                        val parsedCategories = itemsArray.mapNotNull { element ->
                            try {
                                val obj = element.asJsonObject
                                val name = obj.get("name")?.asString ?: return@mapNotNull null
                                val id = obj.get("id")?.asString ?: UUID.randomUUID().toString()
                                val image = obj.get("image_url")?.asString
                                
                                Category(
                                    _id = id,
                                    name = name,
                                    slug = "",
                                    image = image,
                                    icon = null,
                                    parentCategory = null,
                                    group = null,
                                    subCategoryGroupOrder = null
                                )
                            } catch (e: Exception) {
                                null
                            }
                        }
                        
                        _subCategories.value = parsedCategories
                    }
                }
            } catch (e: Exception) {
                AppLogger.debug("Error loading Special Deal layout: ${e.message}")
            }
            
            // fetchProducts function from BasePageManager handles the rest
            selectCategory(null) // triggers fetchProducts internally
            _isLoading.value = false
        }
    }
}
